test/params_api_test.c

```diff
@@ -0,0 +1,362 @@
+            return 0;
+    return 1;
+}
+
+/*
+ * This macro defines a basic test case that sets a param of a specified
+ * integral type, attempts to read the value back and then calls the above
+ * function to verify that the params can be read as other types.
+ *
+ * Since this macro doesn't know the exact sizes of everything, all the real
+ * work is done via byte buffers which are converted to machine order or little
+ * endian for comparisons.  Narrower values are best compared using little
+ * endian because their values and positions don't change.
+ */
+#define TEST_INT(name, type) \
+    static int test_param_##name(int n) \
+    { \
+        type in, out; \
+        unsigned char buf[MAX_LEN], le[MAX_LEN], cmp[sizeof(type)]; \
+        const size_t len = raw_values[n].len >= sizeof(type) \
+                           ? sizeof(type) : raw_values[n].len; \
+        OSSL_PARAM param = OSSL_PARAM_##name("a", NULL); \
+        \
+        memset(buf, 0, sizeof(buf)); \
+        memset(le, 0, sizeof(le)); \
+        copy_be_to_native(buf, raw_values[n].value, len); \
+        swap_copy(le, raw_values[n].value, len); \
+        memcpy(&in, buf, sizeof(in)); \
+        param.buffer = &out; \
+        if (!TEST_true(OSSL_PARAM_set_##name(&param, in))) \
+            return 0; \
+        copy_to_le(cmp, &out, sizeof(out)); \
+        if (!TEST_mem_eq(cmp, len, le, len)) \
+            return 0; \
+        in = 0; \
+        param.buffer = buf; \
+        if (!TEST_true(OSSL_PARAM_get_##name(&param, &in))) \
+            return 0; \
+        copy_to_le(cmp, &in, sizeof(in)); \
+        if (!TEST_mem_eq(cmp, sizeof(in), le, sizeof(in))) \
+            return 0; \
+        param.buffer = &out; \
+        return test_param_type_extra(&param, le, sizeof(type)); \
+    }
+
+/*
+ * The negative test needs to come from outside the macro to avoid
+ * unsigned comparisons that are always true.
+ */
+TEST_INT(int, int)
+TEST_INT(long, long int)
+TEST_INT(uint, unsigned int)
+TEST_INT(ulong, unsigned long int)
+TEST_INT(int32, int32_t)
+TEST_INT(uint32, uint32_t)
+TEST_INT(int64, int64_t)
+TEST_INT(uint64, uint64_t)
+TEST_INT(size_t, size_t)
+
+static int test_param_bignum(int n)
+{
+    unsigned char buf[MAX_LEN], bnbuf[MAX_LEN], le[MAX_LEN];
+    const size_t len = raw_values[n].len;
+    size_t bnsize;
+    BIGNUM *b = NULL, *c = NULL;
+    OSSL_PARAM param = OSSL_PARAM_DEFN("bn", OSSL_PARAM_UNSIGNED_INTEGER,
+                                       NULL, 0, NULL);
+    int ret = 0;
+
+    param.buffer = bnbuf;
+    param.buffer_size = len;
+    param.return_size = &bnsize;
+
+    copy_be_to_native(buf, raw_values[n].value, len);
+    swap_copy(le, raw_values[n].value, len);
+    if (!TEST_ptr(b = BN_bin2bn(raw_values[n].value, (int)len, NULL)))
+        goto err;
+
+    if (!TEST_true(OSSL_PARAM_set_BN(&param, b))
+        || !TEST_mem_eq(bnbuf, bnsize, buf, bnsize)
+        || !TEST_true(OSSL_PARAM_get_BN(&param, &c))
+        || !TEST_BN_eq(b, c))
```

Comment:
So this test is a little interesting, and not quite right.  These things are meant to communicate a bignum to or from "the other end", not to "the same end".  This means that `OSSL_PARAM_set_BN` sets `*param.return_size`, which makes absolute sense for a `OSSL_PARAM` owner, but makes no sense whatsoever to a `OSSL_PARAM` handler (i.e. "the other end" from the `OSSL_PARAM` owner), such as `OSSL_PARAM_get_BN`.

You therefore need to cheat a little, and do a `param.buffer_size = *param.return_size` just before the `OSSL_PARAM_get_BN` call.

Reply:
Note that you have no real need to do this for the native integer tests, 'cause you wouldn't expect `*return_size` to differ from `buffer_size`